(common/i18n): refactor locale generation and update CLDR to v41

gkalpak opened the pull request:
This PR includes some refactorings/improvements to the locale generation process and updates CLDR to version 41.
See individual commits for more details.

Fixes #43301.

##
BTW, you can see a diff of the generated locale files between v39 and v41 [here][1].

[1]: https://github.com/gkalpak/angular/compare/0df46c503b91f0a066907ba2c63863439ed5a6ee..491079b4f09ffa6715296ec4f1f4e637505ccbdf

devversion commented on WORKSPACE:
```diff
@@ -123,12 +123,21 @@ load("@rules_pkg//:deps.bzl", "rules_pkg_dependencies")
+cldr_json_data_repository(
+    name = "cldr_json_data",
+    urls = {
+        "https://github.com/unicode-org/cldr-json/releases/download/%s.0.0/cldr-%s.0.0-json-full.zip" % (2 * (cldr_major_version,)): "a631764b6bb7967fab8cc351aff3ffa3f430a23646899976dd9d65801446def6",
```
Nit: I think it would be more clear to just write out the variable
twice

gkalpak replied:
Done.

devversion commented on packages/common/locales/generate-locales-tool/README.md:
```diff
@@ -59,6 +59,6 @@ If `labelsForDayPeriodsAbbreviated` for example is equal to `labelsForDayPeriods
 ### Where are the CLDR data files stored?
 
-The CLDR data is fetched through a Bazel repository. See the project-level `WORKSPACE` file for more information. The generation tool resolves all locale data through the `@cldr_data` Bazel repository.
+The CLDR data is fetched through Bazel repositories. See the project-level `WORKSPACE` file for more information. The generation tool resolves all locale data through the `@cldr_json_data` and `@cldr_xml_data` Bazel repositories (depending on the data format that is needs).
```
`it needs`?

devversion reviewed (approved):
LGTM, nice work!

dgp1130 commented:
I'll defer to @AndrewKushnir for `fw-i18n` approval as I don't think I understand the runtime behavior enough to give much value here.

AndrewKushnir reviewed (approved):
👍

gkalpak committed:
refactor(common): improve `dayPeriods`-related types in `generate-locales-tool`

Update some TypeScript types in `generate-locales-tool/day-periods.ts`
to more closely reflect the actual shapes of the returned objects.

Examples of the related raw JSON data for the `en` locale can be seen
here:
- [dayPeriods][1]
- [dayPeriodRules][2]

[1]: https://github.com/unicode-org/cldr-json/blob/41.0.0/cldr-json/cldr-dates-full/main/en/ca-gregorian.json#L223-L300
[2]: https://github.com/unicode-org/cldr-json/blob/41.0.0/cldr-json/cldr-core/supplemental/dayPeriods.json#L404-L427

gkalpak committed:
build(common): specify CLDR version to use by `cldr`

While generating locales, two sets of data (both derived from the
[CLDR project][1]) are used:
- JSON data with the [`cldrjs` package][2], which is used for most
  operations.
- XML data with the [`cldr` package][3], which is used for generating
  plural-related data.

The JSON data is brought in from the [unicode-org/cldr-json][4]
repository. Since we control the version of the repository that we use,
we can control the CLDR version that these correspond to.

Previously, however, we used the XML data that were bundled with the
[`cldr` package[3]. As a result, the two sets of data could correspond
to different CLDR versions, resulting in incorrect/inconsistent locales
files.

This commit addresses the problem by utilizing the `load()` method of
the [`cldr` package][3], which allows passing in a custom path to the
CLDR XML data (instead of using the bundled data. This way, we can
ensure that the data used for all operations correspond to the same CLDR
version.

Related discussion: #43301

[1]: https://github.com/unicode-org/cldr
[2]: https://www.npmjs.com/package/cldrjs
[3]: https://www.npmjs.com/package/cldr
[4]: https://github.com/unicode-org/cldr-json

gkalpak committed:
refactor(common): consistently name tooling related to `closure-locale.ts`

Previously, some functions/targets related to the
generation/verification of the `closure-locale.ts` file included
`closure_locales` instead of `closure_locale` in their name, which was
inconsistent and confusing.
Rename all tooling to use `closure_locale` (which more closely matches
the generated file's name).

gkalpak committed:
refactor(common): minor refactoring/simplification in `generate-locales-tool`

Make some minor changes in `generate-locales-tool` to make the code
simpler and/or avoid unnecessary work.

gkalpak committed:
build(common): update CLDR to version 41

Update CLDR, which is used to generate the locales files, to version 41.
Also, make necessary code changes to account for changes in the CLDR
data.

Fixes #43301

gkalpak committed:
fixup! build(common): specify CLDR version to use by `cldr`

gkalpak committed:
fixup! build(common): update CLDR to version 41

gkalpak commented on packages/common/test/i18n/format_date_spec.ts:
```diff
@@ -398,7 +398,7 @@ describe('Format date', () => {
       expect(formatDate(date, 'a', 'hu')).toEqual('de.');
-      expect(formatDate(date, 'a', 'sr')).toEqual('пре подне');
+      expect(formatDate(date, 'a', 'sr')).toEqual('AM');
 
```
AFAICT, this _is_ right, based on the changes to the `sr` locale between [version 39][1] and [version 41][2].

[1]: https://github.com/unicode-org/cldr-json/blob/39.0.0/cldr-json/cldr-dates-full/main/sr/ca-generic.json#L227
[2]: https://github.com/unicode-org/cldr-json/blob/41.0.0/cldr-json/cldr-dates-full/main/sr/ca-generic.json#L227

devversion replied:
sounds good. just wanted to see if you double-checked. thanks!

mary-poppins commented:
You can preview c18ec72 at https://pr45714-c18ec72.ngbuilds.io/.

dylhunn commented:
@gkalpak Is this ready for merge?

AndrewKushnir commented:
@dylhunn I did some additional manual testing to verify that i18n extraction is not affected and everything looks good, so this PR should be ready for merge.
